### PR TITLE
Make use of mocked time to remove test dependency on real time

### PR DESCRIPTION
**Pull request by ZymoticB**
This change introduces a clock concept that enables time to mocked out. This removes the dependency of tests on real time which prevents flakes and also enables tests to be written more accurately around the events that are expected to happen.

Most of the changes are focused in pkg/agent/manager as we saw flakes there. Other changes are mostly to keep the repo compiling. There are remaining places where the globals from the time package are still in use that will need to be cleaned up in follow ups.

There is a behaviour change in the svid rotator ( > -> >=) that makes testing a bit more convenient now that it is perfectly even numbers. This should be extremely minor (microseconds) but its worth mentioning.

There is also a very small timeout change that should be fixed for real by addressing #755

### CODE REVIEW

**Comment by ZymoticB** [comment]
Minor self note, the clock wrapper needs to be moved out of the `test` TLD because it is imported in non-test code.

Managed to find another time related flake in travis. Fixing the server/endpoints/node package now.

**Comment by ZymoticB** [comment]
This is gross https://github.com/spiffe/spire/pull/749/commits/e2e6c6cf922fa4370d45c75534aa83f3d9006744 but I don't see another way. 

I had a flake in travis because (as best as I can tell, unable to repro locally) `time.Until` can be negative here https://github.com/spiffe/spire/blob/master/pkg/server/endpoints/node/limiter.go#L76 if it takes more than 1 millisecond https://github.com/spiffe/spire/blob/master/pkg/server/endpoints/node/limiter_test.go#L22 between the calls to `time.Now()` -- 1 in `context.WithTimeout` and one in `time.Until`. I can't mock out the time as I have in the rest of this PR because of the use of `x/time/rate.Limiter` which internally depends directly on the `time` globals. I don't want to re-write this rate limiter, or wrap in an interface that would have to track `x/time/rate` in this PR. These tests could be improved later I think.

**Comment by azdagron** [comment]
One alternative is to add a test hook channel to the Limiter that is only set by the `newTestLimiter()` function and add a case for a receive from that channel on the `select` inside of `limiter.Limit()` that returns a specific error. In production, the channel would be nil and therefore never considered for selection.

The loop that asserts that tasks aren't blocked would keep the channel open but never send on it and could use a decently high timeout value on the context that prevents false negatives when running under CI but also doesn't block until the `go test` framework times out (10 minutes?) if something goes wrong. 10 seconds, a minute, something like that.

The channel could be closed before testing that the limit is applied. That would cause the error to be returned when the select is hit. It isn't testing the delay value, but then again, the previous test code wasn't either. It does make sure that the code made it to the select.

File: test/util/cert_generation.go [hunk 1]
```diff
@@ -11,32 +11,35 @@ import (
...
 		Subject:            cr.Subject,
 		Issuer:             ca.Subject,
 		PublicKey:          cr.PublicKey,
 		PublicKeyAlgorithm: cr.PublicKeyAlgorithm,
 		Signature:          cr.Signature,
 		SignatureAlgorithm: cr.SignatureAlgorithm,
-		NotBefore:          time.Now(),
-		NotAfter:           time.Now().Add(time.Duration(ttl) * time.Second),
+		NotBefore:          now,
+		NotAfter:           now.Add(time.Duration(ttl) * time.Second),
```
Comment (azdagron) [inline test/util/cert_generation.go]: 💯 getting rid of the two consecutive calls to Now()

File: test/util/cert_generation.go [hunk 2]
```diff
@@ -99,14 +102,14 @@ func Sign(req, parent *x509.Certificate, signerPrivateKey interface{}) (*x509.Ce
...
+func defaultSVIDTemplate(clk clock.Clock) *x509.Certificate {
 	return &x509.Certificate{
 		Subject: pkix.Name{
 			Country:      []string{"US"},
 			Organization: []string{"SPIRE"},
 		},
-		NotBefore: time.Now(),
-		NotAfter:  time.Now().Add(1 * time.Hour),
+		NotBefore: clk.Now(),
+		NotAfter:  clk.Now().Add(1 * time.Hour),
```
Comment (azdagron) [inline test/util/cert_generation.go]: not your bug, but ideally, it seems crisper to calculate NotBefore and NotAfter on the same "now" timestamp.

File: test/util/cert_generation.go [hunk 3]
```diff
@@ -125,8 +128,8 @@ func defaultCATemplate() *x509.Certificate {
 		Subject:               name,
 		Issuer:                name,
 		IsCA:                  true,
-		NotBefore:             time.Now(),
-		NotAfter:              time.Now().Add(1 * time.Hour),
+		NotBefore:             clk.Now(),
+		NotAfter:              clk.Now().Add(1 * time.Hour),
```
Comment (azdagron) [inline test/util/cert_generation.go]: not your bug, but ideally, it seems crisper to calculate NotBefore and NotAfter on the same "now" timestamp.

File: pkg/common/clock/clock.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
+package clock
+
+import (
+	"testing"
```
Comment (azdagron) [inline pkg/common/clock/clock.go]: Feels a little unsanitary for the testing package to end up in production code. Is it worth moving the mock stuff into `test/mocks/mockclock` or `test/fakes/fakeclock` or something?

**Pull request merged**